client-tools/Shell/TelemetricsHandler.cpp

```diff
@@ -0,0 +1,348 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2023 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Puget
+////////////////////////////////////////////////////////////////////////////////
+
+#include "TelemetricsHandler.h"
+
+#include "ApplicationFeatures/HttpEndpointProvider.h"
+#include "ApplicationFeatures/V8PlatformFeature.h"
+#include "ApplicationFeatures/CommunicationFeaturePhase.h"
+#include "Basics/EncodingUtils.h"
+#include "Basics/Exceptions.h"
+#include "Basics/StaticStrings.h"
+#include "Basics/voc-errors.h"
+#include "Endpoint/Endpoint.h"
+#include "Logger/LogMacros.h"
+#include "Shell/ClientFeature.h"
+#include "SimpleHttpClient/GeneralClientConnection.h"
+#include "SimpleHttpClient/SimpleHttpClient.h"
+#include "SimpleHttpClient/SimpleHttpResult.h"
+#include "Ssl/ssl-helper.h"
+#include "Utils/ClientManager.h"
+#include "Rest/GeneralRequest.h"
+#include "V8/v8-deadline.h"
+#include "V8/v8-utils.h"
+
+#include <velocypack/Slice.h>
+
+#include <optional>
+
+using namespace std::chrono;
+using namespace std::chrono_literals;
+
+namespace {
+static std::string getEndpointFromUrl(std::string const& url) {
+  char const* p = url.c_str();
+  char const* e = p + url.size();
+  size_t slashes = 0;
+
+  while (p < e) {
+    if (*p == '?') {
+      // http(s)://example.com?foo=bar
+      return url.substr(0, p - url.c_str());
+    } else if (*p == '/') {
+      if (++slashes == 3) {
+        return url.substr(0, p - url.c_str());
+      }
+    }
+    ++p;
+  }
+
+  return url;
+}
+}  // namespace
+
+namespace arangodb {
+TelemetricsHandler::TelemetricsHandler(ArangoshServer& server,
+                                       bool sendToEndpoint)
+    : _server(server), _sendToEndpoint(sendToEndpoint) {}
+
+TelemetricsHandler::~TelemetricsHandler() {
+  if (_telemetricsThread.joinable()) {
+    _telemetricsThread.join();
+  }
+}
+
+Result TelemetricsHandler::checkHttpResponse(
+    std::unique_ptr<httpclient::SimpleHttpResult> const& response) {
```

Comment:
nice, this function is now side-effects free (except from throwing exceptions), so we can mark it as `const` now!